libsolidity/formal/SMTEncoder.cpp

```diff
@@ -1227,39 +1227,64 @@ pair<smtutil::Expression, smtutil::Expression> SMTEncoder::arithmeticOperation(
 	else
 		intType = TypeProvider::uint256();
 
-	smtutil::Expression valueNoMod(
-		_op == Token::Add ? _left + _right :
-		_op == Token::Sub ? _left - _right :
-		_op == Token::Div ? division(_left, _right, *intType) :
-		_op == Token::Mul ? _left * _right :
-		/*_op == Token::Mod*/ _left % _right
-	);
+	optional<smtutil::Expression> valueUnbounded;
+	if (_op == Token::Add)
+		valueUnbounded = _left + _right;
+	else if (_op == Token::Sub)
+		valueUnbounded = _left - _right;
+	else if (_op == Token::Mul)
+		valueUnbounded = _left * _right;
+	else if (_op == Token::Div)
+		valueUnbounded = division(_left, _right, *intType);
+	else /* _op == Token::Mod */
+		valueUnbounded = _left % _right;
 
 	if (_op == Token::Div || _op == Token::Mod)
+	{
 		m_context.addAssertion(_right != 0);
 
+		// mod and unsigned division never underflow/overflow
+		if (_op == Token::Mod || !intType->isSigned())
+			return {*valueUnbounded, *valueUnbounded};
+
+		// The only case where division overflows is
+		// - type is signed
+		// - LHS is type.min
+		// - RHS is -1
+		// the result is then -(type.min), which wraps back to type.min
+		smtutil::Expression maxLeft = _left == smt::minValue(*intType);
+		smtutil::Expression minusOneRight = _right == -1;
+		smtutil::Expression wrap = smtutil::Expression::ite(maxLeft && minusOneRight, smt::minValue(*intType), *valueUnbounded);
+		return {wrap, *valueUnbounded};
+	}
+
 	auto symbMin = smt::minValue(*intType);
 	auto symbMax = smt::maxValue(*intType);
 
 	smtutil::Expression intValueRange = (0 - symbMin) + symbMax + 1;
+	string suffix = to_string(_operation.id()) + "_" + to_string(m_context.newSlackId());
+	smt::SymbolicIntVariable k(intType, intType, "k_" + suffix, m_context);
+	smt::SymbolicIntVariable m(intType, intType, "m_" + suffix, m_context);
+
+	// To wrap around valueUnbounded in case of overflow or underflow, we replace it with a k, given:
+	// 1. k + m * intValueRange = valueUnbounded
+	// 2. k is in range of the desired integer type
+	auto wrap = k.currentValue();
+	m_context.addAssertion(*valueUnbounded == (k.currentValue() + intValueRange * m.currentValue()));
+	m_context.addAssertion(k.currentValue() >= symbMin);
+	m_context.addAssertion(k.currentValue() <= symbMax);
+
 	auto value = smtutil::Expression::ite(
```

Comment:
We could make this more fine-grained... for unsigned types ``> symbMax`` alone should be enough here.
But maybe we just keep the thought for later, after all the current PRs are all merged... not worth it rechecking everything with that change...

Reply:
Yea, let's do that later. Will add TODO comment

Reply:
Ideally the solvers will just simplify it away anyways based on the type constraints.